Sources/Build/ArtifactsArchiveMetadata.swift

```diff
@@ -0,0 +1,109 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Foundation
+import PackageModel
+import SPMBuildCore
+import TSCBasic
+import TSCUtility
+
+public struct ArtifactsArchiveMetadata: Equatable {
+    public let schemaVersion: String
+    public let artifacts: [String: Artifact]
+
+    public init(schemaVersion: String, artifacts: [String: Artifact]) {
+        self.schemaVersion = schemaVersion
+        self.artifacts = artifacts
+    }
+
+    public struct Artifact: Equatable {
+        let type: ArtifactType
+        let version: String
+        let variants: [Variant]
+
+        public init(type: ArtifactsArchiveMetadata.ArtifactType, version: String, variants: [Variant]) {
+            self.type = type
+            self.version = version
+            self.variants = variants
+        }
+    }
+
+    // In the future we are likely to extend the ArtifactsArchive file format to carry other types of artifacts beyond executables.
+    // Additional fields may be required to support these new artifact types e.g. headers path for libraries.
```

Comment:
It would be good to leave this open enough to perhaps support resource-only artifacts as well. For example, 3d models along with associated textures, or fonts, etc. Of course, it's another matter how to then access that from code — perhaps as resource bundles or something like that.

I know that nothing here prevents non-code artifacts, but if we agree on that as a goal, it would be good to mention specifically that this isn't restricted to just binary code.

Reply:
will add a comment